File: app/src/main/java/at/msd/friehs_bicha/cdcsvparser/MainActivity.java
package at.msd.friehs_bicha.cdcsvparser;

import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

import at.msd.friehs_bicha.cdcsvparser.general.AppModel;


public class MainActivity extends AppCompatActivity {

    private static final int PICKFILE_REQUEST_CODE = 1;
    Context context;
    AppModel appModel;
    File[] files;

    /**
     *sets the buttons and spinner and also fills the global vars files and context the first time
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        context = getApplicationContext();

        //get the elements from the xml.
        Spinner dropdown = findViewById(R.id.spinner_history);
        Button btnParse = findViewById(R.id.btn_parse);
        Button btnHistory = findViewById(R.id.btn_history);

        btnParse.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBtnUploadClick(view);
            }
        });

        updateFiles();

        //disable spinner and history or fill spinner
        if(files.length == 0){
            setHistory("disabled",dropdown,btnHistory);
        }else{
            setSpinner(dropdown);


            btnHistory.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onBtnHistoryClick(view,dropdown);
                }

            });
        }



    }

    /**
     * gets all files from internal file storage and updates it
     * if there are files enable history otherwise disable
     */
    @Override
    protected void onRestart() {
        super.onRestart();

        updateFiles();
        Spinner dropdown = findViewById(R.id.spinner_history);
        Button btnHistory = findViewById(R.id.btn_history);
        if(files.length == 0){
            setHistory("disabled",dropdown,btnHistory);
        }else{
            setHistory("enabled",dropdown,btnHistory);
            btnHistory.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onBtnHistoryClick(view,dropdown);
                }

            });
        }
    }


    /**
     * disables or enbales and fills a spinner
     *
     * @param type "disabled" or "enabled"
     * @param dropdown the Spinner element
     * @param btnHistory the button to de/activate
     */
    private void setHistory(String type, Spinner dropdown,Button btnHistory){
        Resources res = getResources();
        Drawable drawable = res.getDrawable(R.drawable.round_button_layer_list);
        switch(type){
            case "disabled":
                // Disable the button

                btnHistory.setEnabled(false);
                btnHistory.setBackgroundColor(Color.LTGRAY);
                btnHistory.setTextColor(Color.DKGRAY);
                btnHistory.setBackground(drawable);
                //Disable the dropdown
                String[] items = new String[]{"No History"};
                ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item, items);
                dropdown.setAdapter(adapter);
                dropdown.setEnabled(false);
                break;
            case "enabled":
                // Enable the button
                btnHistory.setEnabled(true);
                btnHistory.setTextColor(Color.WHITE);
                btnHistory.setBackground(drawable);
                //Enable the dropdown
                dropdown.setEnabled(true);
                setSpinner(dropdown);
        }

    }

    /**
     * updates the global file array to the newest
     */
    private void updateFiles(){
        // Get the app's internal file directory
        File appDir = getFilesDir();
        // Get a list of all files in the app's internal file directory
        files = appDir.listFiles();
    }

    /**
     * Fills a spinner with parsed names of the global file var
     *
     * @param spinner spinner to fill
     */
    private void setSpinner(Spinner spinner){
        String[] fileNames = new String[files.length];
        SimpleDateFormat sdf = new SimpleDateFormat("M-d-yyyy-hh-mm-ss");
        SimpleDateFormat dateFormat = new SimpleDateFormat("d.M hh:mm");
        String filename;
        Date date;
        for (int i = 0; i < files.length;i++) {
            filename = files[i].getName();
            filename = filename.substring(0, filename.length() - 4);
            try {
                date = sdf.parse(filename);
                filename = dateFormat.format(date);
            } catch (ParseException e) {
                e.printStackTrace();
                // TODO error message
            }
            fileNames[i] = filename;
        }
        ArrayAdapter<String> fileNamesAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item, fileNames);
        spinner.setAdapter(fileNamesAdapter);

    }

    /**
     *Gets the file selected in spinner and reads it to call the parse view
     *
     * @param view the view of the button click
     * @param spinner the spinner to look at
     */
    private void onBtnHistoryClick(View view, Spinner spinner) {
        int position = spinner.getSelectedItemPosition();
        File selectedFile = files[position];

        ArrayList<String> list = getFileContent(selectedFile);

        try {
            appModel = new AppModel(list);
            callParseView();
        }catch (Exception e) {
            CharSequence text = e.getMessage();
            int duration = Toast.LENGTH_SHORT;

            Toast toast = Toast.makeText(context, text, duration);
            toast.show();
        }
    }

    /**
     *is called on successful file select and saves it to storage.
     * Also reads the file and then calls the parse view
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == PICKFILE_REQUEST_CODE && resultCode == RESULT_OK && data != null) {
            // Get the URI of the selected file
            Uri fileUri = data.getData();
            //creat filename with format M-d-y-H-m-s
            SimpleDateFormat dateFormat = new SimpleDateFormat("M-d-y-H-m-s");
            Date now = new Date();
            String time = dateFormat.format(now);
            String filename = time + ".csv";

            ArrayList<String> list = getFileContentFromUri(fileUri);
            try (FileOutputStream fos = context.openFileOutput(filename, Context.MODE_APPEND)) {
                for (String element : list) {
                    fos.write(element.getBytes());
                    fos.write("\n".getBytes());  // add a newline after each element
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            //delete oldest file if alredy 7 files in array
            updateFiles();
            while(files.length > 7){
                files[0].delete();
                updateFiles();
            }

            try {
                appModel = new AppModel(list);
                callParseView();
            }catch (IllegalArgumentException e) {
                context = getApplicationContext();
                CharSequence text = e.getMessage();
                int duration = Toast.LENGTH_SHORT;

                Toast toast = Toast.makeText(context, text, duration);
                toast.show();

            }catch (RuntimeException e){
                Context context = getApplicationContext();
                CharSequence text = e.getMessage();
                int duration = Toast.LENGTH_SHORT;

                Toast toast = Toast.makeText(context, text, duration);
                toast.show();
                callParseView();
            }


        }
    }

    private void callParseView() {
        Intent intent = new Intent(MainActivity.this, ParseActivity.class);
        intent.putExtra("AppModel", appModel);
        startActivity(intent);
    }

    /**
     * start action to let the user select a file
     */
    public void onBtnUploadClick(View view){
        // Create an Intent object to allow the user to select a file
        Intent chooseFile = new Intent(Intent.ACTION_GET_CONTENT);

        // Set the type of file that the user can select
        chooseFile.setType("*/*");

        // Start the activity to let the user select a file
        startActivityForResult(chooseFile, PICKFILE_REQUEST_CODE);
    }

    /**
     * gets the file from the url then reads it and returns it
     *
     * @param uri url to file
     * @return the content of the file in a ArrayList<Sting>
     */
    public ArrayList<String> getFileContentFromUri(Uri uri){
        ArrayList<String> fileContents = new ArrayList<>();
        try {
            // Get the ContentResolver for the current context.
            ContentResolver resolver = getContentResolver();

            // Open an InputStream for the file represented by the Uri.
            InputStream inputStream = resolver.openInputStream(uri);

            // Create a BufferedReader to read the file contents.
            BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));

            // Read the file line by line and add each line to the fileContents list.
            String line;
            while ((line = reader.readLine()) != null) {
                fileContents.add(line);
            }

            // Close the BufferedReader and InputStream.
            reader.close();
            inputStream.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return fileContents;
    }

    /**
     * reads a file and returns it
     *
     * @param file a file
     * @return the content of the file in a ArrayList<Sting>
     */
    public ArrayList<String> getFileContent(File file){
        ArrayList<String> fileContents = new ArrayList<>();
        try {

            // Create a BufferedReader to read the file contents.
            BufferedReader reader = new BufferedReader(new FileReader(file));

            // Read the file line by line and add each line to the fileContents list.
            String line;
            while ((line = reader.readLine()) != null) {
                fileContents.add(line);
            }

            // Close the BufferedReader and InputStream.
            reader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return fileContents;
    }

}

File: app/src/main/java/at/msd/friehs_bicha/cdcsvparser/MainActivity.kt
package at.msd.friehs_bicha.cdcsvparser

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.res.ResourcesCompat
import at.msd.friehs_bicha.cdcsvparser.core.CoreService
import at.msd.friehs_bicha.cdcsvparser.logging.FileLog
import at.msd.friehs_bicha.cdcsvparser.util.Benchmarker
import at.msd.friehs_bicha.cdcsvparser.util.FileUtil
import at.msd.friehs_bicha.cdcsvparser.util.PreferenceHelper
import com.google.firebase.auth.FirebaseAuth
import java.io.*
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity() {
    var context: Context? = null
    var files: Array<File>? = null
    var user = FirebaseAuth.getInstance().currentUser
    private lateinit var progressDialog: Dialog

    // SAF file picker (ActivityResult API): no storage permissions are
    // needed for ACTION_GET_CONTENT, so the old READ_EXTERNAL_STORAGE /
    // READ_MEDIA_* permission dance is gone.
    private val pickFile =
        registerForActivityResult(ActivityResultContracts.GetContent()) { uri: Uri? ->
            if (uri != null) onFileSelected(uri) else hideProgressDialog()
        }

    companion object {
        // Single source for the history file name pattern (write + parse).
        // Older versions wrote "M-d-y-H-m-s" which is still tolerated on read
        // failures (raw file name is shown instead).
        const val HISTORY_FILE_PATTERN = "yyyy-MM-dd-HH-mm-ss"
    }


    /**
     * sets the buttons and spinner and also fills the global vars files and context the first time
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        context = applicationContext

        user = FirebaseAuth.getInstance().currentUser

        //get the elements from the xml.
        val dropdown = findViewById<Spinner>(R.id.spinner_history)
        val btnParse = findViewById<Button>(R.id.btn_parse)
        val btnHistory = findViewById<Button>(R.id.btn_history)
        val btnLoadFromDB = findViewById<Button>(R.id.btn_loadFromDb)
        if (user == null) {
            btnLoadFromDB.visibility = View.GONE
        } else {
            btnLoadFromDB.visibility = View.VISIBLE
        }
        btnParse.setOnClickListener { onBtnUploadClick() }
        btnLoadFromDB.setOnClickListener { loadFromFireBaseDB() }
        settingsButton()
        updateFiles()

        //disable spinner and history or fill spinner
        if (files!!.isEmpty()) {
            setHistory("disabled", dropdown, btnHistory)
        } else {
            setSpinner(dropdown)
            btnHistory.setOnClickListener { onBtnHistoryClick(dropdown) }
        }

        if (intent.hasExtra("fastStart")) {
            fastStart()
        }
    }

    private fun fastStart() {
        val intent = Intent(this@MainActivity, ParseActivity::class.java)
        startActivity(intent)
    }


    /**
     * gets all files from internal file storage and updates it
     * if there are files enable history otherwise disable
     */
    override fun onRestart() {
        super.onRestart()
        updateFiles()
        val dropdown = findViewById<Spinner>(R.id.spinner_history)
        val btnHistory = findViewById<Button>(R.id.btn_history)
        if (files!!.isEmpty()) {
            setHistory("disabled", dropdown, btnHistory)
        } else {
            setHistory("enabled", dropdown, btnHistory)
            btnHistory.setOnClickListener { view: View? -> onBtnHistoryClick(dropdown) }
        }
    }


    /**
     * disables or enbales and fills a spinner
     *
     * @param type       "disabled" or "enabled"
     * @param dropdown   the Spinner element
     * @param btnHistory the button to de/activate
     */
    private fun setHistory(type: String, dropdown: Spinner, btnHistory: Button) {
        val drawable = ResourcesCompat.getDrawable(resources, R.drawable.round_button_layer_list, null)
        when (type) {
            "disabled" -> {
                // Disable the button
                btnHistory.isEnabled = false
                btnHistory.setBackgroundColor(Color.LTGRAY)
                btnHistory.setTextColor(Color.DKGRAY)
                btnHistory.background = drawable
                //Disable the dropdown
                val items = arrayOf(resources.getString(R.string.no_history))
                val adapter =
                    ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, items)
                dropdown.adapter = adapter
                dropdown.isEnabled = false
            }

            "enabled" -> {
                // Enable the button
                btnHistory.isEnabled = true
                btnHistory.setTextColor(Color.WHITE)
                btnHistory.background = drawable
                //Enable the dropdown
                dropdown.isEnabled = true
                setSpinner(dropdown)
            }
        }
    }

    /**
     * updates the global file array to the newest
     */
    private fun updateFiles() {
        // Get the app's internal file directory
        val appDir = filesDir
        // Get a list of all files in the app's internal file directory
        files = appDir.listFiles() as Array<File>
        files = files!!.filter { it.name.endsWith(".csv") }.toTypedArray()
    }

    /**
     * Fills a spinner with parsed names of the global file var
     *
     * @param spinner spinner to fill
     */
    @SuppressLint("SimpleDateFormat")
    private fun setSpinner(spinner: Spinner) {
        val fileNames = ArrayList<String>()
        // Files imported by this version use HISTORY_FILE_PATTERN. Files from
        // older versions are simply shown with their raw file name.
        val sdf = SimpleDateFormat(HISTORY_FILE_PATTERN)
        val dateFormat = SimpleDateFormat("d.M HH:mm")
        var filename: String
        for (f in files!!) {
            if (!f.isFile || !f.name.endsWith(".csv")) continue
            filename = f.name
            filename = filename.substring(0, filename.length - 4)
            try {
                filename = dateFormat.format(sdf.parse(filename))
            } catch (e: ParseException) {
                e.printStackTrace()
                FileLog.e("MainActivity", "setSpinner: Date parse error: $e")
            }
            fileNames.add(filename)
        }
        val fileNamesAdapter =
            ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, fileNames)
        spinner.adapter = fileNamesAdapter
    }

    /**
     * Gets the file selected in spinner and reads it to call the parse view
     *
     * @param spinner the spinner to look at
     */
    private fun onBtnHistoryClick(spinner: Spinner) {
        showProgressDialog()
        val position = spinner.selectedItemPosition
        val selectedFile = files!![position]
        val list = FileUtil.getFileContent(selectedFile)
        try {
            Benchmarker.start()
            CoreService.startServiceWithData(list, PreferenceHelper.getSelectedType(this).ordinal)
            callParseView()
        } catch (e: Exception) {
            hideProgressDialog()
            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * is called on successful file select and saves it to storage.
     * Also reads the file and then calls the parse view
     */
    private fun onFileSelected(fileUri: Uri) {
        showProgressDialog()
        // create filename with format yyyy-MM-dd-HH-mm-ss (see HISTORY_FILE_PATTERN)
        val dateFormat = SimpleDateFormat(HISTORY_FILE_PATTERN)
        val now = Date()
        val time = dateFormat.format(now)
        val filename = "$time.csv"
        val list = FileUtil.getFileContentFromUri(this, fileUri)
        try {
            applicationContext.openFileOutput(filename, MODE_PRIVATE).use { fos ->
                for (element in list) {
                    fos.write(element.toByteArray())
                    fos.write("\n".toByteArray()) // add a newline after each element
                }
            }
        } catch (e: IOException) {
            FileLog.e("MainActivity", ":  Error while writing to file : $e")
        }
        //delete oldest file if already 7 files in array
        updateFiles()
        while (files!!.size > 7) {
            files!![0].delete()
            updateFiles()
        }
        Benchmarker.start()
        CoreService.startServiceWithData(
            list,
            PreferenceHelper.getSelectedType(this).ordinal
        )
        callParseView()
    }

    private fun callParseView(saveToDB: Boolean = true) {
        if (saveToDB) {
            CoreService.saveDataToFirebase()
        }
        val intent = Intent(this@MainActivity, ParseActivity::class.java)
        hideProgressDialog()
        startActivity(intent)
    }


    /**
     * start action to let the user select a file (SAF, no permissions)
     */
    private fun onBtnUploadClick() {
        showProgressDialog()
        pickFile.launch("*/*")
    }


    private fun settingsButton() {
        val settingsButton = findViewById<Button>(R.id.settings_button)
        settingsButton.setOnClickListener {
            val intent = Intent(this@MainActivity, SettingsActivity::class.java)
            startActivity(intent)
        }
    }


    private fun loadFromFireBaseDB() {
        showProgressDialog()
//start ACTION_START_SERVICE_WITH_FIREBASE_DATA
        val intent = Intent(this@MainActivity, CoreService::class.java)
        intent.action = CoreService.ACTION_START_SERVICE_WITH_FIREBASE_DATA
        startService(intent)

        callParseView(saveToDB = false)

    }

    fun showProgressDialog() {
        progressDialog = Dialog(this)
        progressDialog.setContentView(R.layout.progress_icon)
        progressDialog.setCancelable(false)
        progressDialog.setCanceledOnTouchOutside(false)
        progressDialog.show()
    }

    fun hideProgressDialog() {
        progressDialog.dismiss()
    }

}